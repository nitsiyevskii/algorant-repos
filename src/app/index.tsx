import { View, Text, StyleSheet } from 'react-native';

export default function Index() {
  return (
    <View style={styles.container}>
      <Text style={styles.text}>Home - Repository List</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff',
  },
  text: {
    fontSize: 18,
  },
});
