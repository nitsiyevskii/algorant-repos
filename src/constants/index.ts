export * from './github';
export * from './api';
export * from './filters';

