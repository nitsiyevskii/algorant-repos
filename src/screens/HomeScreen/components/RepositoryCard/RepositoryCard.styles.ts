import { createStyles } from "src/utils/createStyles";

export const useStyles = createStyles((theme) => ({
  container: {
    backgroundColor: theme.colors.background.white,
    borderRadius: 12,
    padding: 16,
    marginBottom: 12,
    borderWidth: 1,
    borderColor: theme.colors.separator.primary,
  },
  header: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "flex-start",
    marginBottom: 8,
  },
  titleContainer: {
    flex: 1,
    marginRight: 12,
  },
  title: {
    fontSize: theme.fonts.size.lg,
    fontWeight: theme.fonts.weight.semibold,
    color: theme.colors.text.primary,
    marginBottom: 4,
  },
  owner: {
    fontSize: theme.fonts.size.sm,
    color: theme.colors.text.secondary,
  },
  description: {
    fontSize: theme.fonts.size.md,
    color: theme.colors.text.secondary,
    marginBottom: 12,
    lineHeight: 20,
  },
  footer: {
    flexDirection: "row",
    alignItems: "center",
    gap: 16,
  },
  stat: {
    flexDirection: "row",
    alignItems: "center",
    gap: 4,
  },
  statText: {
    fontSize: theme.fonts.size.sm,
    color: theme.colors.text.tertiary,
  },
  languageDot: {
    width: 12,
    height: 12,
    borderRadius: 6,
  },
}));

