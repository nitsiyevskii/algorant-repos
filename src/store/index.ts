import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

import repositoriesReducer from './slices/repositoriesSlice';
import filtersReducer from './slices/filtersSlice';
import repositoryDetailsReducer from './slices/repositoryDetailsSlice';

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
  whitelist: ['repositories'],
};

const rootReducer = combineReducers({
  repositories: repositoriesReducer,
  filters: filtersReducer,
  repositoryDetails: repositoryDetailsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

