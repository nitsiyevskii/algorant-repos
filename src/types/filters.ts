export interface RepositoryFilters {
  organizations: string[];
  languages: string[];
  stars: {
    min: number | null;
    max: number | null;
  };
  forks: {
    min: number | null;
    max: number | null;
  };
}

export const DEFAULT_FILTERS: RepositoryFilters = {
  organizations: [],
  languages: [],
  stars: {
    min: null,
    max: null,
  },
  forks: {
    min: null,
    max: null,
  },
};

