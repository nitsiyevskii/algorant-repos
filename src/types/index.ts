export type { Theme, ThemeColors, ThemeFonts } from './theme';

