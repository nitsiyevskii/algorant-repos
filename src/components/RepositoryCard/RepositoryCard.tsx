import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { router } from 'expo-router';
import { useTheme } from 'src/hooks/useTheme';
import { FavoriteButton } from 'src/components/FavoriteButton/FavoriteButton';
import { GitHubRepository } from 'src/types/github';
import { useStyles } from './RepositoryCard.styles';

interface RepositoryCardProps {
  repository: GitHubRepository;
  isFavorite: boolean;
  onToggleFavorite: () => void;
}

export const RepositoryCard: React.FC<RepositoryCardProps> = ({
  repository,
  isFavorite,
  onToggleFavorite,
}) => {
  const theme = useTheme();
  const styles = useStyles();

  const handlePress = () => {
    router.push(`/repository/${encodeURIComponent(repository.full_name)}`);
  };

  return (
    <TouchableOpacity
      style={styles.container}
      onPress={handlePress}
      activeOpacity={0.7}
    >
      <View style={styles.header}>
        <View style={styles.titleContainer}>
          <Text style={styles.title} numberOfLines={1}>
            {repository.name}
          </Text>
          <Text style={styles.owner} numberOfLines={1}>
            {repository.owner.login}
          </Text>
        </View>
        <FavoriteButton isSelected={isFavorite} onPress={onToggleFavorite} />
      </View>

      {repository.description && (
        <Text style={styles.description} numberOfLines={2}>
          {repository.description}
        </Text>
      )}

      <View style={styles.footer}>
        <View style={styles.stat}>
          <Ionicons name="star" size={16} color={theme.colors.text.tertiary} />
          <Text style={styles.statText}>{repository.stargazers_count}</Text>
        </View>

        {repository.language && (
          <View style={styles.stat}>
            <View
              style={[styles.languageDot, { backgroundColor: theme.colors.core.primary }]}
            />
            <Text style={styles.statText}>{repository.language}</Text>
          </View>
        )}

        <View style={styles.stat}>
          <Ionicons name="git-branch" size={16} color={theme.colors.text.tertiary} />
          <Text style={styles.statText}>{repository.forks_count}</Text>
        </View>
      </View>
    </TouchableOpacity>
  );
};

