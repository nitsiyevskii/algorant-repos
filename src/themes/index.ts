import { Theme } from 'src/types/theme';
import { colors } from './colors';
import { fonts } from './fonts';

export const lightTheme: Theme = {
  colors,
  fonts,
};

export const darkTheme: Theme = {
  colors: {
    ...colors,
    text: {
      primary: '#FFFFFF',
      secondary: '#E9EAE9',
      tertiary: '#BEC0C4',
      label: '#FFFFFF',
      error: '#FF453A',
      success: '#32D74B',
    },
    background: {
      white: '#000000',
      black: '#FFFFFF',
      primary: '#1C1C1E',
      secondary: '#2C2C2E',
      tertiary: '#3A3A3C',
    },
    input: {
      background: '#1C1C1E',
      border: '#3A3A3C',
      text: '#FFFFFF',
      placeholder: '#8E8E93',
      focused: '#0A84FF',
      error: '#FF453A',
    },
    button: {
      primary: '#0A84FF',
      secondary: '#5E5CE6',
      disabled: '#3A3A3C',
      text: '#FFFFFF',
    },
    iconButton: {
      background: '#2C2C2E',
      icon: '#FFFFFF',
      disabled: '#3A3A3C',
    },
    core: {
      primary: '#0A84FF',
      secondary: '#5E5CE6',
      accent: '#FF9F0A',
      error: '#FF453A',
      success: '#32D74B',
      warning: '#FF9F0A',
      info: '#64D2FF',
    },
    modal: {
      overlay: 'rgba(0, 0, 0, 0.75)',
      background: '#1C1C1E',
    },
    separator: {
      primary: '#38383A',
      secondary: '#48484A',
    },
    textDisabled: {
      primary: '#3A3A3C',
      secondary: '#48484A',
    },
  },
  fonts,
};

export { colors } from './colors';
export { fonts } from './fonts';

export type { Theme, ThemeColors, ThemeFonts } from 'src/types/theme';
