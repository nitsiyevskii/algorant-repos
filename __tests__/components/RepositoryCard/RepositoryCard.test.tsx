import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { router } from 'expo-router';
import { RepositoryCard } from 'src/components/RepositoryCard/RepositoryCard';
import { GitHubRepository } from 'src/types';

const mockRepository: GitHubRepository = {
  id: 1,
  name: 'test-repo',
  full_name: 'algorand/test-repo',
  owner: {
    login: 'algorand',
    avatar_url: 'https://test.com/avatar.png',
    html_url: 'https://github.com/algorand',
  },
  html_url: 'https://github.com/algorand/test-repo',
  description: 'Test repository description',
  fork: false,
  created_at: '2023-01-01T00:00:00Z',
  updated_at: '2023-01-02T00:00:00Z',
  pushed_at: '2023-01-02T00:00:00Z',
  stargazers_count: 100,
  watchers_count: 50,
  language: 'TypeScript',
  forks_count: 10,
  open_issues_count: 5,
  topics: ['blockchain', 'algorand'],
  visibility: 'public',
  default_branch: 'main',
};

describe('RepositoryCard', () => {
  const mockOnToggleFavorite = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders repository information correctly', () => {
    const { getByText } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    expect(getByText('test-repo')).toBeTruthy();
    expect(getByText('algorand')).toBeTruthy();
    expect(getByText('Test repository description')).toBeTruthy();
  });

  it('displays repository statistics', () => {
    const { getByText } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    expect(getByText('100')).toBeTruthy();
    expect(getByText('TypeScript')).toBeTruthy();
    expect(getByText('10')).toBeTruthy();
  });

  it('renders favorite button with correct state', () => {
    const { getByTestId } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={true}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    const favoriteButton = getByTestId('favorite-button');
    expect(favoriteButton).toBeTruthy();
  });

  it('calls onToggleFavorite when favorite button is pressed', () => {
    const { getByTestId } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    const favoriteButton = getByTestId('favorite-button');
    fireEvent.press(favoriteButton);

    expect(mockOnToggleFavorite).toHaveBeenCalledTimes(1);
  });

  it('navigates to repository details when pressed', () => {
    const { getByText } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    const card = getByText('test-repo').parent?.parent?.parent;
    if (card) {
      fireEvent.press(card);
      expect(router.push).toHaveBeenCalledWith('/repository/algorand/test-repo');
    }
  });

  it('renders without description', () => {
    const repoWithoutDescription = {
      ...mockRepository,
      description: null,
    };

    const { queryByText } = render(
      <RepositoryCard
        repository={repoWithoutDescription}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    expect(queryByText('Test repository description')).toBeNull();
  });

  it('renders without language', () => {
    const repoWithoutLanguage = {
      ...mockRepository,
      language: null,
    };

    const { queryByText } = render(
      <RepositoryCard
        repository={repoWithoutLanguage}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    expect(queryByText('TypeScript')).toBeNull();
  });

  it('matches snapshot when not favorite', () => {
    const { toJSON } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when favorite', () => {
    const { toJSON } = render(
      <RepositoryCard
        repository={mockRepository}
        isFavorite={true}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('truncates long repository names', () => {
    const repoWithLongName = {
      ...mockRepository,
      name: 'very-long-repository-name-that-should-be-truncated',
    };

    const { getByText } = render(
      <RepositoryCard
        repository={repoWithLongName}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    const nameElement = getByText('very-long-repository-name-that-should-be-truncated');
    expect(nameElement.props.numberOfLines).toBe(1);
  });

  it('truncates long descriptions', () => {
    const repoWithLongDescription = {
      ...mockRepository,
      description: 'This is a very long description that should be truncated to two lines maximum to ensure proper display',
    };

    const { getByText } = render(
      <RepositoryCard
        repository={repoWithLongDescription}
        isFavorite={false}
        onToggleFavorite={mockOnToggleFavorite}
      />
    );

    const descriptionElement = getByText('This is a very long description that should be truncated to two lines maximum to ensure proper display');
    expect(descriptionElement.props.numberOfLines).toBe(2);
  });
});

