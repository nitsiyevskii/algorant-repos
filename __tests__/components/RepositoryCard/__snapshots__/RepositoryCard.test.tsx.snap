// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryCard matches snapshot when favorite 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#C6C6C8",
      "borderRadius": 12,
      "borderWidth": 1,
      "marginBottom": 12,
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 8,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "marginRight": 12,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#000000",
            "fontSize": 18,
            "fontWeight": "600",
            "marginBottom": 4,
          }
        }
      >
        test-repo
      </Text>
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#3C3C43",
            "fontSize": 14,
          }
        }
      >
        algorand
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F2F2F7",
          "borderRadius": 20,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
      testID="favorite-button"
    >
      <Text
        color="#FF3B30"
        name="heart"
        size={24}
        testID="icon"
      >
        heart
      </Text>
    </View>
  </View>
  <Text
    numberOfLines={2}
    style={
      {
        "color": "#3C3C43",
        "fontSize": 16,
        "lineHeight": 20,
        "marginBottom": 12,
      }
    }
  >
    Test repository description
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        color="#8E8E93"
        name="star"
        size={16}
        testID="icon"
      >
        star
      </Text>
      <Text
        style={
          {
            "color": "#8E8E93",
            "fontSize": 14,
          }
        }
      >
        100
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "width": 12,
            },
            {
              "backgroundColor": "#007AFF",
            },
          ]
        }
      />
      <Text
        style={
          {
            "color": "#8E8E93",
            "fontSize": 14,
          }
        }
      >
        TypeScript
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        color="#8E8E93"
        name="git-branch"
        size={16}
        testID="icon"
      >
        git-branch
      </Text>
      <Text
        style={
          {
            "color": "#8E8E93",
            "fontSize": 14,
          }
        }
      >
        10
      </Text>
    </View>
  </View>
</View>
`;

exports[`RepositoryCard matches snapshot when not favorite 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#C6C6C8",
      "borderRadius": 12,
      "borderWidth": 1,
      "marginBottom": 12,
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 8,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "marginRight": 12,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#000000",
            "fontSize": 18,
            "fontWeight": "600",
            "marginBottom": 4,
          }
        }
      >
        test-repo
      </Text>
      <Text
        numberOfLines={1}
        style={
          {
            "color": "#3C3C43",
            "fontSize": 14,
          }
        }
      >
        algorand
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#F2F2F7",
          "borderRadius": 20,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
      testID="favorite-button"
    >
      <Text
        color="#000000"
        name="heart-outline"
        size={24}
        testID="icon"
      >
        heart-outline
      </Text>
    </View>
  </View>
  <Text
    numberOfLines={2}
    style={
      {
        "color": "#3C3C43",
        "fontSize": 16,
        "lineHeight": 20,
        "marginBottom": 12,
      }
    }
  >
    Test repository description
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 16,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        color="#8E8E93"
        name="star"
        size={16}
        testID="icon"
      >
        star
      </Text>
      <Text
        style={
          {
            "color": "#8E8E93",
            "fontSize": 14,
          }
        }
      >
        100
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "width": 12,
            },
            {
              "backgroundColor": "#007AFF",
            },
          ]
        }
      />
      <Text
        style={
          {
            "color": "#8E8E93",
            "fontSize": 14,
          }
        }
      >
        TypeScript
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        color="#8E8E93"
        name="git-branch"
        size={16}
        testID="icon"
      >
        git-branch
      </Text>
      <Text
        style={
          {
            "color": "#8E8E93",
            "fontSize": 14,
          }
        }
      >
        10
      </Text>
    </View>
  </View>
</View>
`;
